### PR TITLE
SOLR-16340 Provide a cluster_id label for all metrics exposed by the Solr Prometheus Exporter

### PR DESCRIPTION
**Pull request by janhoy**
https://issues.apache.org/jira/browse/SOLR-16340

* Added label `cluster_id` on all metrics from exporter
* Added filter (template) for `cluster_id` in grafana board
* Defaults to `base_url` or `zk_host`, but can be overridden by `--cluster-id foo` or `--id foo` command-line or env.var `CLUSTER_ID`.

~Lacks a test for the actual setting of `clusterId` and argument parsing, since existing junit tests do not invoke the exporter through shell~. **Edit**: Docker test now validates env.var, and thus also cmdline arg parsing.

I keep `clusterId` as a static var on `SolrExporter.java` to avoid passing it through all kinds of intermediate method calls to `SolrScraper`, but there is probably a more elegant and more testable way..

### CODE REVIEW

File: solr/prometheus-exporter/bin/solr-exporter.cmd [hunk 1]
```diff
@@ -82,6 +82,7 @@ IF NOT "%SCRAPE_INTERVAL%"=="" set EXPORTER_ARGS=%EXPORTER_ARGS% -s %SCRAPE_INTE
 IF NOT "%NUM_THREADS%"=="" set EXPORTER_ARGS=%EXPORTER_ARGS% -n %NUM_THREADS%
 IF NOT "%ZK_HOST%"=="" set EXPORTER_ARGS=%EXPORTER_ARGS% -z %ZK_HOST%
 IF NOT "%SOLR_URL%"=="" set EXPORTER_ARGS=%EXPORTER_ARGS% -b %SOLR_URL%
+IF NOT "%CLUSTER_ID%"=="" set EXPORTER_ARGS=%EXPORTER_ARGS% -i %CLUSTER_ID%
```
Comment (janhoy) [inline solr/prometheus-exporter/bin/solr-exporter.cmd:85]: Is this space-safe wrt quoting?

**Comment by HoustonPutman** [comment]
@janhoy I added a PR that removed the use of the static variable, and also added zk_host and base_url labels to the scrapingErrors metric.

Feel free to revert if you don't like it!

**Comment by janhoy** [comment]
Nice improvement. Good point about labeling exporter’s own metrics! Should grafana board display some exporter metrics ?

**Comment by HoustonPutman** [comment]
> Should grafana board display some exporter metrics ?

Yeah probably. I haven't looked over the grafana changes here, but will hopefully have time to help with those later this week!

File: solr/prometheus-exporter/bin/solr-exporter.cmd (hunk 1 above)
Comment (janhoy) [inline solr/prometheus-exporter/bin/solr-exporter.cmd:85]:
```suggestion
IF NOT "%CLUSTER_ID%"=="" set EXPORTER_ARGS=%EXPORTER_ARGS% -i "%CLUSTER_ID%"
```
Think normal quoting will do...

**Comment by janhoy** [comment]
Committed quoting of ClusterID in windows script, and resolved CHANGES conflict. Looking forward to final reviews.

**Comment by HoustonPutman** [comment]
> Looking forward to final reviews.

Going to spin this up today with a few clusters and poke around!

**Comment by HoustonPutman** [comment]
Ok the dashboard looks good! Just a few changes from my end, mainly that "Shard Leaders", "Replica State History" and "QPS (Collection)" Did not specify the cluster_id, and "QPS (Collection)" did not group by it either. I will push a change with this, but the only problem is the default cluster_id. If it's a long zk-connection-string, then this will probably become problematic and make the UI somewhat unusable. (since the cluster_id will be extremely long).

Maybe instead we use a default such as "main", and require users to override it? I get the convenience of using a zk-connection-string, since that is by-default a 1-1 mapping with cluster_id, but maybe we require users to provide the id?

Would appreciate your opinion here.

Also sidenote, in the future maybe we can change this to a graph for multiple clusters? (Not in this PR though)
Screen Shot 2022-09-06 at 2 55 27 PM

**Comment by HoustonPutman** [comment]
Here is an example of using zkConnectionString as cluster_id:

Screen Shot 2022-09-06 at 4 11 00 PM

**Comment by HoustonPutman** [comment]
> So another option could perhaps be defaulting to a hash of zkHost or solrUrl, e.g. the first 10 chars of a sha hash

I love it.

I fixed a small error, but I think we should be good to go now!

**Review by HoustonPutman — APPROVED** [review]

**Comment by janhoy** [comment]
Ah, that was why the Docker test failed. Nice to have end-to-end tests this way.

I’ll proceed with merge tomorrow..

**Pull request merged**